src/plots/cartesian/ordered_categories.js

```diff
@@ -0,0 +1,33 @@
+/**
+* Copyright 2012-2016, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var d3 = require('d3');
+
+
+/**
+ * TODO add documentation
+ */
+module.exports = function orderedCategories(axisLetter, categorymode, categorylist, data) {
+
+    return categorymode === 'array' ?
+
+        // just return a copy of the specified array ...
+        categorylist.slice() :
+
+        // ... or take the union of all encountered tick keys and sort them as specified
+        // (could be simplified with lodash-fp or ramda)
+        [].concat.apply([], data.map(function(d) {return d[axisLetter];}))
+            .filter(function(element, index, array) {return index === array.indexOf(element);})
+            .sort(({
```

Comment:
We like using for loops in situations. Although they may be not as nice to look at than these functional patterns, they are much [faster](https://jsperf.com/array-loop-vs-filter). :racehorse: